File: backend/agent/tools/serp.py
"""
Phantom — SERP API Tool (Job Discovery)

Uses Bright Data SERP API to discover job postings from a plain English query.
This is the DISCOVERY tool — it finds jobs the user never had to search for.

Input: search query string (e.g., "software engineer fintech remote")
Output: list of SERPResult objects with job URLs, titles, and snippets

Bright Data SERP API docs: https://docs.brightdata.com/scraping-automation/serp-api
"""

import os
import logging
import httpx
from urllib.parse import quote_plus
from typing import Optional

from models.schemas import SERPResult

logger = logging.getLogger(__name__)

# Bright Data SERP API endpoint
SERP_API_URL = "https://api.brightdata.com/request"


async def fetch_serp_results(
    query: str,
    max_results: int = 10,
    **kwargs,
) -> list[SERPResult]:
    """
    Discover job postings via Bright Data SERP API.

    Constructs a targeted Google search query to find job listings
    across major job boards, then returns structured results.

    Args:
        query: Plain English job search query (e.g., "software engineer fintech remote")
        max_results: Maximum number of results to return (default 10)

    Returns:
        List of SERPResult objects containing title, URL, and snippet for each job found.
        Returns empty list if the API call fails.
    """
    api_key = os.getenv("BRIGHT_DATA_API_KEY")
    zone = os.getenv("BRIGHT_DATA_SERP_ZONE", "serp")

    if not api_key:
        logger.error("BRIGHT_DATA_API_KEY not set")
        return []

    # Build a search query targeting job boards.
    # URL-encode with quote_plus so spaces → "+" and special chars are escaped.
    # Bright Data's API validates the URL and rejects raw spaces.
    search_query = f"{query} jobs hiring now"
    encoded_query = quote_plus(search_query)
    search_url = (
        f"https://www.google.com/search"
        f"?q={encoded_query}"
        f"&num={max_results}"
        f"&hl=en"
    )

    payload = {
        "zone": zone,
        "url": search_url,
        "format": "json",   # Required by Bright Data Web Access API
        "brd_json": True,   # Return parsed SERP JSON (Google organic results)
    }

    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {api_key}",
    }

    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.post(
                SERP_API_URL,
                json=payload,
                headers=headers,
            )
            response.raise_for_status()
            data = response.json()

        results = []
        # Parse organic search results from the SERP response
        organic = data.get("organic", data.get("results", []))
        if isinstance(organic, list):
            for item in organic[:max_results]:
                result = SERPResult(
                    title=item.get("title", ""),
                    url=item.get("link", item.get("url", "")),
                    snippet=item.get("description", item.get("snippet", "")),
                    source="SERP API",
                )
                if result.url:  # Only include results with valid URLs
                    results.append(result)

        logger.info(f"SERP API returned {len(results)} results for query: {query}")
        return results

    except httpx.HTTPStatusError as e:
        logger.error(f"SERP API HTTP error: {e.response.status_code} - {e.response.text}")
        return []
    except httpx.RequestError as e:
        logger.error(f"SERP API request error: {e}")
        return []
    except Exception as e:
        logger.error(f"SERP API unexpected error: {e}")
        return []


async def extract_job_info_from_serp(results: list[SERPResult]) -> list[dict]:
    """
    Extract structured job information from raw SERP results.

    Parses titles and snippets to identify company names, locations,
    and job titles from the search results.

    Args:
        results: List of raw SERPResult objects from fetch_serp_results

    Returns:
        List of dicts with keys: job_title, company, location, url
    """
    jobs = []
    for result in results:
        # Extract basic info — the title often contains "Job Title - Company - Location"
        title_parts = result.title.split(" - ")
        job_info = {
            "job_title": title_parts[0].strip() if title_parts else result.title,
            "company": title_parts[1].strip() if len(title_parts) > 1 else "Unknown",
            "location": title_parts[2].strip() if len(title_parts) > 2 else "Not specified",
            "url": result.url,
            "snippet": result.snippet,
        }
        jobs.append(job_info)

    return jobs

File: src/lib/agents/serp.ts
/**
 * SERP Agent — Discovers job postings via Bright Data SERP API (Google search).
 *
 * This is Phase 1 of the pipeline: it takes a plain English query, searches
 * Google for job postings, and returns structured results with titles, URLs,
 * and snippets. The results are then enriched by the other 3 agents.
 *
 * @module agents/serp
 */
import { getEnv } from "@/lib/config/env";
import { BRIGHT_DATA_API_URL } from "@/lib/data";
import { parseGoogleSerpHtml } from "@/lib/parsers/serp-html";
import type { SERPResult } from "@/lib/types";

export async function fetchSerpResults(query: string, maxResults = 10): Promise<SERPResult[]> {
  const env = getEnv();
  const searchQuery = `${query} jobs hiring now`;
  const encodedQuery = encodeURIComponent(searchQuery);
  const searchUrl = `https://www.google.com/search?q=${encodedQuery}&num=${maxResults}&hl=en`;

  const payload = {
    zone: env.BRIGHT_DATA_SERP_ZONE,
    url: searchUrl,
  };

  const headers = {
    "Content-Type": "application/json",
    Authorization: `Bearer ${env.BRIGHT_DATA_API_KEY}`,
  };

  try {
    const response = await fetch(BRIGHT_DATA_API_URL, {
      method: "POST",
      headers,
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(30_000),
    });

    if (!response.ok) {
      console.error(`SERP API HTTP ${response.status}: ${await response.text().catch(() => "")}`);
      return [];
    }

    const html = await response.text();
    const parsed = parseGoogleSerpHtml(html, maxResults);

    const results: SERPResult[] = parsed.map((item) => ({
      title: item.title,
      url: item.url,
      snippet: item.snippet,
      source: "SERP API",
    }));

    console.log(`SERP API returned ${results.length} results for: ${query}`);
    return results;
  } catch (err) {
    console.error(`SERP API error: ${err}`);
    return [];
  }
}

/**
 * Runs multiple query variations through SERP in parallel.
 * Returns the combined (non-deduplicated) results from all variations.
 * Deduplication happens in the relevance scorer.
 */
export async function fetchMultiSerpResults(
  queryVariations: string[],
  maxResultsPerQuery = 10
): Promise<SERPResult[]> {
  const results = await Promise.all(
    queryVariations.map((q) => fetchSerpResults(q, maxResultsPerQuery))
  );
  const combined = results.flat();
  console.log(
    `Multi-SERP: ${queryVariations.length} variations → ${combined.length} total results ` +
    `(before dedup)`
  );
  return combined;
}

export async function extractJobInfoFromSerp(results: SERPResult[]): Promise<{
  job_title: string;
  company: string;
  location: string;
  url: string;
  snippet: string;
}[]> {
  return results.map((result) => {
    const parsed = parseJobTitle(result.title, result.url);
    return {
      job_title: parsed.job_title,
      company: parsed.company,
      location: parsed.location,
      url: result.url,
      snippet: result.snippet,
    };
  });
}

/**
 * Parses job title strings from various job board formats:
 * - Indeed:    "Software Engineer - Google - Mountain View, CA"
 * - LinkedIn:  "Software Engineer | Google | Mountain View, CA"
 * - Glassdoor: "Software Engineer Jobs at Google | Glassdoor"
 * - Generic:   "Software Engineer at Google (Remote)"
 *
 * Falls back to extracting company from URL domain if all patterns fail.
 */
/** @internal Exported for testing */
export function parseJobTitle(title: string, url: string): {
  job_title: string;
  company: string;
  location: string;
} {
  // Pattern 1: "Title | Company | Location" (LinkedIn style)
  if (title.includes(" | ")) {
    const parts = title.split(" | ").map((s) => s.trim());
    // Filter out site names like "LinkedIn", "Glassdoor", "Indeed"
    const filtered = parts.filter(
      (p) => !/(linkedin|glassdoor|indeed|ziprecruiter|monster|dice)\b/i.test(p)
    );
    if (filtered.length >= 2) {
      return {
        job_title: filtered[0],
        company: filtered[1],
        location: filtered[2] || "Not specified",
      };
    }
  }

  // Pattern 2: "Title at Company" or "Title @ Company"
  const atMatch = title.match(/^(.+?)\s+(?:at|@)\s+(.+?)(?:\s*[-–|]\s*(.+))?$/i);
  if (atMatch) {
    return {
      job_title: atMatch[1].trim(),
      company: atMatch[2].trim(),
      location: atMatch[3]?.trim() || "Not specified",
    };
  }

  // Pattern 3: "Title - Company - Location" (Indeed style, most common)
  const dashParts = title.split(/\s*[-–—]\s*/);
  if (dashParts.length >= 2) {
    // Filter out site names from the parts
    const filtered = dashParts.filter(
      (p) => !/(linkedin|glassdoor|indeed|ziprecruiter|monster|dice|jobs?$)\b/i.test(p.trim())
    );
    if (filtered.length >= 2) {
      return {
        job_title: filtered[0].trim(),
        company: filtered[1].trim(),
        location: filtered[2]?.trim() || "Not specified",
      };
    }
  }

  // Fallback: try to extract company from URL domain
  const company = extractCompanyFromUrl(url);
  return {
    job_title: title.trim(),
    company,
    location: "Not specified",
  };
}

/** Extracts a company name guess from the URL domain (e.g., "stripe.com" → "Stripe") */
/** @internal Exported for testing */
export function extractCompanyFromUrl(url: string): string {
  try {
    const hostname = new URL(url).hostname;
    // Skip job board domains — they're not the company
    const jobBoards = ["indeed.com", "linkedin.com", "glassdoor.com", "ziprecruiter.com", "monster.com", "dice.com", "lever.co", "greenhouse.io", "workday.com"];
    if (jobBoards.some((jb) => hostname.includes(jb))) return "Unknown";
    // Take the main domain part and capitalize it
    const domain = hostname.replace(/^www\./, "").split(".")[0];
    return domain.charAt(0).toUpperCase() + domain.slice(1);
  } catch {
    return "Unknown";
  }
}
